_posts/2020-01-13-mixed-precision-training/mixed_precision.Rmd

```diff
@@ -0,0 +1,292 @@
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Source: https://images.nvidia.com/content/volta-architecture/pdf/volta-architecture-whitepaper.pdf."}
+knitr::include_graphics("images/fma.png")
+```
+
+The operation takes place on 4x4 matrices; multiplications happen on 16-bit operands while the final result could be 16-bit or 32-bit.
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Source: https://images.nvidia.com/content/volta-architecture/pdf/volta-architecture-whitepaper.pdf"}
+knitr::include_graphics("images/fma2.png")
+```
+
+We can see how this is immediately relevant to the operations involved in deep learning; the details, however, are [not necessarily clear](https://www.anandtech.com/show/12673/titan-v-deep-learning-deep-dive/3). 
+
+Leaving those internals to the experts, we now proceed to the actual experiment.
+
+## Experiments
+
+### Dataset
+
+With their 28x28px / 32x32px sized images, neither MNIST nor CIFAR seemed particularly suited to challenge the GPU. Instead, we chose [_Imagenette_](https://github.com/fastai/imagenette), the "little ImageNet" created by the _fast.ai_ folks, consisting of 10 classes: _tench, English springer, cassette player, chain saw, church, French horn, garbage truck, gas pump, golf ball,_ and _parachute_. ^[We do hope usage is allowed even in case we can't produce the required "corny inauthentic French accent".] Here are a few examples, taken from the 320px version:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Examples of the 10 classes of Imagenette."}
+knitr::include_graphics("images/results.jpg")
+```
+
+These images have been resized - keeping the aspect ratio - such that the larger dimension has length 320px. As part of preprocessing, we'll further resize to 256x256px, to work with a nice power of 2. ^[As per the [documentation](https://tensorflow.rstudio.com/guide/keras/mixed_precision/), the number of filters in a convolutional layer should be a multiple of 8; however, taking this additional measure couldn't possibly hurt.]
+
+The dataset may conveniently be obtained via using [tfds](), the R interface to TensorFlow Datasets. 
+
+```{r}
+library(keras)
+# needs version 2.1
+library(tensorflow)
+library(tfdatasets)
+# available from github: devtools::install_github("rstudio/tfds")
+library(tfds)
+
+# to use TensorFlow Datasets, we need the Python backend
+# normally, just use tfds::install_tfds for this
+# as of this writing though, we need a nightly build of TensorFlow Datasets
+# envname should refer to whatever environment you run TensorFlow in
+reticulate::py_install("tfds-nightly", envname = "r-reticulate") 
+
+# on first execution, this downloads the dataset
+imagenette <- tfds_load("imagenette/320px")
+
+# extract train and test parts
+train <- imagenette$train
+test <- imagenette$validation
+
+# batch size for the initial run
+batch_size <- 32
```

Comment:
Have you tried a bigger batch size? 32 might be too small to see any differences in performance.

Reply:
thanks for commenting! I was trying batch sizes between and 32 and 512, for that reason...

Reply:
Oops I failed at RTFA, you're right :)